fluent-bit-init-process.go

```diff
@@ -0,0 +1,402 @@
+package main
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/sirupsen/logrus"
+)
+
+// static paths
+const (
+	s3FileFolderPath           = "fluent-bit-init-s3-files"
+	mainConfigFilePath         = "fluent-bit-init.conf"
+	originalMainConfigFilePath = "/fluent-bit/etc/fluent-bit.conf"
+	invokerFilePath            = "fluent-bit-invoker.sh"
+)
+
+var (
+	// default Fluent Bit command
+	fluentBitCommand = "exec /fluent-bit/bin/fluent-bit -e /fluent-bit/firehose.so -e /fluent-bit/cloudwatch.so -e /fluent-bit/kinesis.so"
+
+	// global s3 client and flag
+	s3Client      *s3.S3
+	exists3Client bool = false
+
+	// global ecs metadata region
+	metadataReigon string = ""
+)
+
+// HTTPClient interface
+type HTTPClient interface {
+	Get(url string) (*http.Response, error)
+}
+
+// S3Downloader interface
+type S3Downloader interface {
+	Download(w io.WriterAt, input *s3.GetObjectInput, options ...func(*s3manager.Downloader)) (int64, error)
+}
+
+type ECSTaskMetadata struct {
+	AWS_REGION          string `json:"AWSRegion"`      // AWS region
+	ECS_CLUSTER         string `json:"Cluster"`        // Cluster name
+	ECS_TASK_ARN        string `json:"TaskARN"`        // Task ARN
+	ECS_TASK_ID         string `json:"TaskID"`         // Task Id
+	ECS_FAMILY          string `json:"Family"`         // Task family
+	ECS_LAUNCH_TYPE     string `json:"LaunchType"`     // Task launch type, will be an empty string if container agent is under version 1.45.0
+	ECS_REVISION        string `json:"Revision"`       // Revision number
+	ECS_TASK_DEFINITION string `json:"TaskDefinition"` // TaskDefinition = "family:revision"
+}
+
+// create the HTTP Client
+func createHTTPClient() HTTPClient {
+	client := &http.Client{}
+	return client
+}
+
+// get ECS Task Metadata via endpoint V4
+func getECSTaskMetadata(httpClient HTTPClient) ECSTaskMetadata {
+
+	var metadata ECSTaskMetadata
+
+	ecsTaskMetadataEndpointV4 := os.Getenv("ECS_CONTAINER_METADATA_URI_V4")
+	if ecsTaskMetadataEndpointV4 == "" {
+		logrus.Warnln("[FluentBit Init Process] Unable to get ECS Metadata")
+		return metadata
+	}
+
+	res, err := httpClient.Get(ecsTaskMetadataEndpointV4 + "/task")
+	if err != nil {
+		logrus.Warnln(err)
+		logrus.Warnln("[FluentBit Init Process] Failed to get ECS Metadata via HTTP Get")
```

Comment:
For same level, why don't we merge them like `logrus.Warnln("[FluentBit Init Process] Failed to get ECS Metadata via HTTP Get: %s", err)`

Reply:
and I think it should be error level. If we cannot get it, we should directly return right? The logic should be the same as the file I gave you..

Reply:
Yeah, for the warning message, will merge them.

For the second question, I remember that you said inject ECS Metadata should be an optional.. So when running on ECS, init process can get the endpoint, but if run this image locally, customers can also use it normally just can't use metadata as env vars..

Reply:
yeah but I don't think error level will exit the code immediately?

Reply:
because you will directly return here if you are not in ECS and cannot get the endpoint: https://github.com/aws/aws-for-fluent-bit/pull/379/files/8c9902211f1d542a7ba372724898dbb62aca94bd#diff-a24b896daae00708d98524f520e1d75af05aea6586bd9e70568e44197652681e. So in this line, if we cannot get metadata from the endpoint, we should be in an error right?